remove FDB entires of BridgePort before removing the BridgePort

madhanmellanox opened the pull request:
<!--
Please make sure you have read and understood the contribution guildlines:
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

1. Make sure your commit includes a signature generted with `git commit -s`
2. Make sure your commit title follows the correct format: [component]: description
3. Make sure your commit message contains enough details about the change and related tests
4. Make sure your pull request adds related reviewers, asignees, labels

Please also provide the following information in this pull request:
-->

**What I did**
To fix an issue when removing bridgeport, I introduced code to remove FDB entries of the bridgeport before calling SAI API remove BridgePort.

**Why I did it**
I introduced this fix because by directly removing the bridgeport without flushing the bridge port's FDB entries, the bridge port removal was failing, since it had uncleaned ASIC DB FDB entries.

**How I verified it**
I verified it by sending traffic to the vlan port and creating a FDB entry for the vlan port. By, then when I remove the bridge port member from the VLAN port, the FDB entry was removed as well as bridge port is removed from VLAN membership. 

**Details if related**
The fix is introduced in PortsOrch.cpp

Commit:
remove FDB entires of BridgePort before removing the BridgePort

keboliu commented on orchagent/portsorch.cpp:
```diff
@@ -3431,6 +3462,14 @@ bool PortsOrch::removeBridgePort(Port &port)
     }
 
+    //Flush the FDB entires corresponding to the port
+    if (!flushFDBEntries(port.m_bridge_port_id))
```
one port can be added to multi VLAN, when we flush the mac we should only flush the mac address learned in VLAN which is moving out? what do you think?

madhanmellanox replied:
Let us say, port Ethernet0 is a member of VLAN 2. The FDB entry that is learnt for the port is in VLAN 1. But, once we try to remove Ethernet 0 from being member of VLAN2, looks like we have to remove FDB entries of the bridge port corresponding to VLAN 2 as well as VLAN 1. Is that right?

madhanmellanox replied:
@kebo, please ignore my above comment. removeBridgePort() member function is called only if the port is no more part of any VLANs. So, the call to flush FDB entries with no VLAN information is right. It will flush all the Fdb etries corresponding to the port.

liat-grozovik commented on orchagent/portsorch.cpp:
```diff
@@ -3130,6 +3130,37 @@ void PortsOrch::doTask(Consumer &consumer)
+    if (SAI_NULL_OBJECT_ID == bridge_port_id)
+    {
+        SWSS_LOG_WARN("Couldn't flush FDB. Bridge port OID: 0x%" PRIx64,
+                      bridge_port_id);
```
if it is null object id how can you print it?

madhanmellanox replied:
changed to port.m_alias.c_str() instead of bridge_port_id. But, printing bridge_port_id was part of fdborch.cpp. That is how it was done there. Now, since we move to printing port.m_alias.c_str(), it will display the port name say, Ethernet120 for example.

liat-grozovik commented on orchagent/portsorch.cpp:
```diff
@@ -3130,6 +3130,37 @@ void PortsOrch::doTask(Consumer &consumer)
+    attr.value.oid = bridge_port_id;
+    attrs.push_back(attr);
+
+    SWSS_LOG_INFO("Flushing FDB bridge_port_oid: 0x%" PRIx64, bridge_port_id);
```
what are the values of bridge_port_id? a magic number on the sw know of the user also know it?
should we provide a more clear info to the user?

madhanmellanox replied:
yes, now port.m_alias.c_str() is getting printed instead.

liat-grozovik commented on orchagent/portsorch.cpp:
```diff
@@ -3130,6 +3130,37 @@ void PortsOrch::doTask(Consumer &consumer)
+    if (SAI_STATUS_SUCCESS != rv)
+    {
+        SWSS_LOG_ERROR("Flushing FDB failed. rv:%d", rv);
+	return false;
```
indentation problem

madhanmellanox replied:
addressed in the next commit.

liat-grozovik commented on orchagent/portsorch.cpp:
```diff
@@ -3431,6 +3462,14 @@ bool PortsOrch::removeBridgePort(Port &port)
+    if (!flushFDBEntries(port.m_bridge_port_id))
+    {	    
+        SWSS_LOG_ERROR("Failed to flush FDB entries for the port %s",
+                port.m_alias.c_str());
```
remove 'the'.
why it is different from the bridge_port_id you displayed before?
are you referring to the interface? what is port from user perspective?

madhanmellanox replied:
Removed 'the' word. Yes, I am referring to the interface.

liat-grozovik commented on orchagent/portsorch.h:
```diff
@@ -127,6 +127,7 @@ class PortsOrch : public Orch, public Subject
     bool addSubPort(Port &port, const string &alias, const bool &adminUp = true, const uint32_t &mtu = 0);
     bool removeSubPort(const string &alias);
     void getLagMember(Port &lag, vector<Port> &portv);
+    bool flushFDBEntries(sai_object_id_t);
```
rename to flushFdbEntries

madhanmellanox replied:
addressed in next commit.

Commit:
fixed function naming and other print related information

Commit:
fixed indentation issue

madhanmellanox commented:
@liat-grozovik can you please approve this PR? I have addressed all your comments.

liat-grozovik commented:
It is approved
The only Q is if other places in the code uses port and not interfaces when referring to fdb

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: madhanmellanox <notifications@github.com>
Sent: Sunday, September 27, 2020 8:34:50 AM
To: madhanmellanox/sonic-swss <sonic-swss@noreply.github.com>
Cc: Liat Grozovik <liatg@nvidia.com>; Mention <mention@noreply.github.com>
Subject: Re: [madhanmellanox/sonic-swss] remove FDB entires of BridgePort before removing the BridgePort (#3)


@liat-grozovik<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fliat-grozovik&data=02%7C01%7Cliatg%40nvidia.com%7C97e1704cd685471f7e2b08d862a70fcd%7C43083d15727340c1b7db39efd9ccc17a%7C0%7C0%7C637367816957150447&sdata=K6x3cjlASZcCi8QtkOdzmtr%2FxcYWGo3DpEal3kVuaDU%3D&reserved=0> can you please approve this PR? I have addressed all your comments.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmadhanmellanox%2Fsonic-swss%2Fpull%2F3%23issuecomment-699588637&data=02%7C01%7Cliatg%40nvidia.com%7C97e1704cd685471f7e2b08d862a70fcd%7C43083d15727340c1b7db39efd9ccc17a%7C0%7C0%7C637367816957150447&sdata=UGTBpDLPqzP2OVkZIqTGeAHjHNSJ8xY%2F7T3r%2B%2FiDe3I%3D&reserved=0>, or unsubscribe<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKTABA5UDP7KWXBTMNFTWRDSH3FHVANCNFSM4RXPK3QQ&data=02%7C01%7Cliatg%40nvidia.com%7C97e1704cd685471f7e2b08d862a70fcd%7C43083d15727340c1b7db39efd9ccc17a%7C0%7C0%7C637367816957160439&sdata=JYKnjExfinBuFlmHW8xPQnWiqFzv8VSa4U7ebYzeY88%3D&reserved=0>.

madhanmellanox commented:
> It is approved The only Q is if other places in the code uses port and not interfaces when referring to fdb Get Outlook for iOS<https://aka.ms/o0ukef>
> […](#)
> ________________________________ From: madhanmellanox <notifications@github.com> Sent: Sunday, September 27, 2020 8:34:50 AM To: madhanmellanox/sonic-swss <sonic-swss@noreply.github.com> Cc: Liat Grozovik <liatg@nvidia.com>; Mention <mention@noreply.github.com> Subject: Re: [madhanmellanox/sonic-swss] remove FDB entires of BridgePort before removing the BridgePort (#3) @liat-grozovik<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fliat-grozovik&data=02%7C01%7Cliatg%40nvidia.com%7C97e1704cd685471f7e2b08d862a70fcd%7C43083d15727340c1b7db39efd9ccc17a%7C0%7C0%7C637367816957150447&sdata=K6x3cjlASZcCi8QtkOdzmtr%2FxcYWGo3DpEal3kVuaDU%3D&reserved=0> can you please approve this PR? I have addressed all your comments. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmadhanmellanox%2Fsonic-swss%2Fpull%2F3%23issuecomment-699588637&data=02%7C01%7Cliatg%40nvidia.com%7C97e1704cd685471f7e2b08d862a70fcd%7C43083d15727340c1b7db39efd9ccc17a%7C0%7C0%7C637367816957150447&sdata=UGTBpDLPqzP2OVkZIqTGeAHjHNSJ8xY%2F7T3r%2B%2FiDe3I%3D&reserved=0>, or unsubscribe<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKTABA5UDP7KWXBTMNFTWRDSH3FHVANCNFSM4RXPK3QQ&data=02%7C01%7Cliatg%40nvidia.com%7C97e1704cd685471f7e2b08d862a70fcd%7C43083d15727340c1b7db39efd9ccc17a%7C0%7C0%7C637367816957160439&sdata=JYKnjExfinBuFlmHW8xPQnWiqFzv8VSa4U7ebYzeY88%3D&reserved=0>.

The only place fdb entry is used in portsorch.cpp is this, it is using Port for printing. But in fdborch.cpp, it always uses m_bridge_port_id for printing.

Commit:
Fixing function name in portsorch.h

Commit:
added function call to flushFdbEntries() also in removeDefaultBridgePorts() function

liat-grozovik commented on orchagent/portsorch.cpp:
```diff
@@ -444,6 +444,16 @@ void PortsOrch::removeDefaultBridgePorts()
         if (attr.value.s32 == SAI_BRIDGE_PORT_TYPE_PORT)
         {
+            Port port;
+	    getPortByBridgePortId(bridge_port_list[i], port);
```
indentation problem

Commit:
fixed indentation issues

Commit:
instead of a new member function for portsorch used fdborch member function

Commit:
moved flushFDBEntries member function of fdborch from private to public

madhanmellanox commented:
This is an internal PR, closing this PR since the corresponding Azure PR got merged
